zebra/rt_netlink.c

```diff
@@ -2686,7 +2686,7 @@ static ssize_t netlink_neigh_update_msg_encode(
 
 	if (datalen < sizeof(*req))
 		return 0;
-	memset(req, 0, datalen);
+	memset(req, 0, sizeof(*req));
```

Comment:
I think this is wrong, you are only `memset`ting the header and skipping the payload. Is this intentional? Am I missing something?

Reply:
I thought this was ok - no need to do the work of touching all the bytes of the buffer; the bytes that are going to be sent are going to be set to ... something useful?

Reply:
I don't like to gamble with random values popping in the middle of unset headers/payload and I think that may give opening to vulnerabilities (e.g. reading previous stack/heap values somehow).

Like I said: this works, but I'd assume not setting some field means `0` which in this case it is not covered for payload. Also I think this case is harder for static analysers to catch "uninitialised memory".

Reply:
oh man, I so dislike disagreeing with you, Rafael! but I don't think we always NULL-out 8K of data buffer before sending ... a couple of hundred (or a couple of dozen) ?

Reply:
That's ok, it is a non blocking comment I already approved the PR.

Have you measured the impact of `memset`ting few bytes vs 8k buffers? Are you worried we are using too big buffers then?

I still think that this is the wrong place to attempt to get performance and the correct one is to avoid syscalls/context switches which is the direction we are already going to with batching.

Reply:
This is intentional. We build all the messages that we send to the kernel this way. We only clear those two structures in the `req` struct. And with the message batching, this would be even a bigger problem - the data buffer can be bigger than 8K. If potential vulnerabilities are the serious problem then we can solve this by passing already zeroed buffer.

Reply:
>  ... If potential vulnerabilities are the serious problem then we can solve this by passing already zeroed buffer.

@xThaid that works and will look better too! 

We could go a step further by standardizing the netlink encoding functions to always receive a buffer, so we'd get less code: (1) by zeroing buffers at declaration and (2) reducing the differences (e.g. when to use `sizeof(req)` or `datalen`).

But this is already outside of the scope of the PR.